File: Design-Patterns-Using-Cpp/Decorator/DecoratorPattern.cpp
#include <iostream>

// Component
class Component {
public:
    virtual void operation() = 0;
    virtual ~Component() = default;
};

// ConcreteComponent
class ConcreteComponent : public Component {
public:
    void operation() override {
        std::cout << "ConcreteComponent operation." << std::endl;
    }
};

// Decorator
class Decorator : public Component {
private:
    Component* component;

public:
    Decorator(Component* component) : component(component) {}

    void operation() override {
        if (component) {
            component->operation();
        }
    }

    virtual ~Decorator() {
        delete component;
    }
};

// ConcreteDecorator1
class ConcreteDecorator1 : public Decorator {
public:
    ConcreteDecorator1(Component* component) : Decorator(component) {}

    void operation() override {
        std::cout << "ConcreteDecorator1 operation start." << std::endl;
        Decorator::operation();
        std::cout << "ConcreteDecorator1 operation end." << std::endl;
    }
};

// ConcreteDecorator2
class ConcreteDecorator2 : public Decorator {
public:
    ConcreteDecorator2(Component* component) : Decorator(component) {}

    void operation() override {
        std::cout << "ConcreteDecorator2 operation start." << std::endl;
        Decorator::operation();
        std::cout << "ConcreteDecorator2 operation end." << std::endl;
    }
};

// Client code
int main() {
    Component* component = new ConcreteComponent();
    Decorator* decorator1 = new ConcreteDecorator1(component);
    Decorator* decorator2 = new ConcreteDecorator2(decorator1);

    decorator2->operation();

    // Clean up dynamically allocated memory
    delete decorator2;

    return 0;
}

/*
ConcreteDecorator2 operation start.
ConcreteDecorator1 operation start.
ConcreteComponent operation.
ConcreteDecorator1 operation end.
ConcreteDecorator2 operation end.
*/

File: Design-Patterns-Using-Go/Decorator/DecoratorPattern.go
package main

import "fmt"

// Component is the interface for the base component
type Component interface {
	Operation()
}

// ConcreteComponent is the concrete implementation of Component
type ConcreteComponent struct{}

// Operation implements the Operation method for ConcreteComponent
func (c *ConcreteComponent) Operation() {
	fmt.Println("ConcreteComponent operation.")
}

// Decorator is the interface for the decorator
type Decorator interface {
	Operation()
}

// BaseDecorator provides default implementation for the Operation method
type BaseDecorator struct {
	component Component
}

// Operation implements the Operation method for BaseDecorator
func (d *BaseDecorator) Operation() {
	d.component.Operation()
}

// ConcreteDecorator1 is a concrete decorator
type ConcreteDecorator1 struct {
	BaseDecorator
}

// Operation implements the Operation method for ConcreteDecorator1
func (d *ConcreteDecorator1) Operation() {
	fmt.Println("ConcreteDecorator1 operation start.")
	d.BaseDecorator.Operation()
	fmt.Println("ConcreteDecorator1 operation end.")
}

// ConcreteDecorator2 is another concrete decorator
type ConcreteDecorator2 struct {
	BaseDecorator
}

// Operation implements the Operation method for ConcreteDecorator2
func (d *ConcreteDecorator2) Operation() {
	fmt.Println("ConcreteDecorator2 operation start.")
	d.BaseDecorator.Operation()
	fmt.Println("ConcreteDecorator2 operation end.")
}

// Client code
func main() {
	component := &ConcreteComponent{}
	decorator1 := &ConcreteDecorator1{BaseDecorator{component}}
	decorator2 := &ConcreteDecorator2{BaseDecorator{decorator1}}
	decorator2.Operation()
}

/*
ConcreteDecorator2 operation start.
ConcreteDecorator1 operation start.
ConcreteComponent operation.
ConcreteDecorator1 operation end.
ConcreteDecorator2 operation end.
*/
